server/src/main/java/org/elasticsearch/snapshots/package-info.java

```diff
@@ -0,0 +1,112 @@
+ * (signaled by the shard's snapshot state being {@code INIT}).
+ * For those local primary shards found in state {@code INIT}) the snapshot process
+ * of writing the shard's data files to the snapshot's {@link org.elasticsearch.repositories.Repository} is executed.
+ * Once the snapshot execution finishes for a shard an {@code UpdateIndexShardSnapshotStatusRequest} is sent to the master node signaling
+ * either status {@code SUCCESS} or {@code FAILED}.
+ * The master node will then update a shard's state in the snapshots {@code SnapshotsInProgress.Entry} whenever
+ * it receives such a {@code UpdateIndexShardSnapshotStatusRequest}.</li>
+ *
+ * <li>If as a result of the received status update requests, all shards in the cluster-state are in a completed state, i.e are marked as
+ * either {@code SUCCESS}, {@code FAILED} or {@code MISSING}, the {@code SnapshotShardsService}
+ * will update the state of the {@code Entry} itself and mark it as {@code SUCCESS}.
+ * At the same time {@link org.elasticsearch.snapshots.SnapshotsService#endSnapshot} is executed, writing the metadata necessary to
+ * finalize the snapshot in the repository to the repository.</li>
+ *
+ * <li>After writing the final metadata to the repository, a cluster-state update to remove the snapshot from the repository is submitted
+ * and the removal of the snapshot's {@code SnapshotsInProgress.Entry} from the cluster-state completes the snapshot process.</li>
+ * </ol>
+ *
+ * <h1>Deleting a Snapshot</h1>
+ *
+ * <p>Deleting a snapshot can take the form of either simply deleting it from the repository or (if it has not completed yet) aborting it
+ * and subsequently deleting it from the repository.</p>
+ *
+ * <h2>Aborting a Snapshot</h2>
+ *
+ * <ol>
+ * <li>Aborting a snapshot starts by updating the state of the snapshot's {@code SnapshotsInProgress.Entry} to {@code ABORTED}.</li>
+ *
+ * <li>The snapshot's state change to {@code ABORTED} in cluster state is then picked up by the {@code SnapshotShardsService} on all nodes.
+ * Those nodes that have shard snapshot actions for the snapshot assigned to them, will abort them and notify master about the shards
+ * snapshot status accordingly. If the shard snapshot action completed or was in state {@code FINALIZE} when the abort was registered by
+ * the {@code SnapshotShardsService}, then the shard's state will be reported to master as {@code SUCCESS}.
+ * Otherwise, it will be reported as {@code FAILED}.</li>
+ *
+ * <li>Once all the shards are reported to master as either {@code SUCCESS} or
+ * {@code FAILED} the {@code SnapshotsService} on the master will finish the snapshot process as all shard's
+ * states are now completed and hence the snapshot can be completed as explained in point 4 of the snapshot creation section above.</li>
+ * </ol>
+ *
+ * <h2>Deleting a Snapshot from a Repository</h2>
+ *
+ * <ol>
+ * <li>Assuming there are no entries in the cluster-state's {@code SnapshotsInProgress}, deleting a snapshot
```

Comment:
It'd also be interesting to hear how we delete a snapshot after aborting it, i.e. what if this `Assuming` is not true?

Reply:
The is covered in the paragraph above, that's why I put `Aborting a Snapshot` as a subtopic of `Deleting a Snapshot`. Should I move things around a bit to make it clearer?

Reply:
The detail I felt we were missing is how we achieve that "after" relationship. Aborting a snapshot isn't atomic, it requires acks from all the shards first, so how do we know to delete the snapshot after aborting it? It looks like we add a listener on the _current_ master so all is good as long as we don't elect a new master first.

Reply:
Yea true, the delete after abort will not happen if we have a master failover. Then the abort will go through clean but no delete will happen subsequently. That's something we could maybe fix I guess but I'm not sure it's worth the risk (at least at this point in time).